tests/unit/test_charm.py

```diff
@@ -308,3 +333,52 @@ def test_pebble_ready(harness: Harness, patch: SpringBootPatch):
     harness.begin_with_initial_hooks()
     harness.container_pebble_ready("spring-boot-app")
     assert isinstance(harness.model.unit.status, ActiveStatus)
+
+
+def test_datasource(harness: Harness, patch: SpringBootPatch):
+    """
+    arrange: provide a simulated Spring Boot application image.
+    act: start the charm and a database. Integrates the db to the charm.
+    assert: test datasource output of the charm when integrated with a db or not.
+        Datasource contains the data needed to spring-boot
+        to establish a connection to the data provider.
+    """
+    patch.start(
+        {"spring-boot-app": OCIImageMock.builder().add_file("/app/test.jar", b"").build()},
+    )
+    harness.begin_with_initial_hooks()
+    harness.container_pebble_ready("spring-boot-app")
+    assert isinstance(harness.model.unit.status, ActiveStatus)
+    assert harness.charm._datasource() == {"username": "", "password": "", "url": ""}
+
+    mysql_relation_id: int = harness.add_relation("mysql", "mysql-k8s")
+
+    # Simulate sharing the credentials of a new created database.
+    # https://github.com/canonical/charm-relation-interfaces/blob/main/interfaces/mysql_client/v0/README.md#example
+    mysql_host = "mysql"
+    mysql_port = 3306
+
+    # Test without the database name provided by the provider
+    # This implies that the default "spring-boot" name was used
+    rel_data = {
+        "endpoints": f"{mysql_host}:{mysql_port}",
+        "password": "test-password",
+        "username": "test-username",
+    }
+    harness.update_relation_data(mysql_relation_id, "mysql-k8s", rel_data)
+    assert harness.charm._datasource() == {
+        "password": rel_data["password"],
+        "username": rel_data["username"],
+        "url": f"jdbc:mysql://{mysql_host}:{mysql_port}/spring-boot",
+    }
+
+    # Now we add a database name given by the provider
```

Comment:
Should these be multiple tests rather than checking multiple things in the same tests? There is a trade-off for setup complexity, although setup can be done using fixtures. This could potentially be parametrized as well: https://docs.pytest.org/en/6.2.x/parametrize.html

Reply:
Changed to use parametrize